crates/nu-protocol/src/value/duration.rs

```diff
@@ -0,0 +1,239 @@
+    pub fn normalize(&mut self) {
+        let overflow_days = self.lsb.num_days();
+        if overflow_days == 0 {
+            return;
+        };
+
+        // carry days out of lsb to msb
+        self.lsb = self.lsb - Duration::days(overflow_days);
+        self.msb = self.msb + self.msb.with_days(overflow_days as i32);
+
+        // (hopefully) normalize month, week, day in msb
+        let end_date = self.base.date_naive() + self.msb;
+        self.msb = RelativeDuration::from_duration_between(self.base.date_naive(), end_date);
+    }
+
+    /// Returns struct with "normalized" fields
+    pub fn as_struct(&mut self) -> NuDurationStruct {
+        self.normalize();
+
+        // reduce sub-day units first
+
+        let (seconds, fraction) = divmod(self.lsb.num_nanoseconds().unwrap_or(0), 1_000_000_000);
+        let (minutes, seconds) = divmod(seconds, 60);
+        let (hours, minutes) = divmod(minutes, 60);
+        let (days, hours) = divmod(hours, 24);
+
+        assert_eq!(days, 0, "excess overflow days");
+
+        // assume normalize rationalized msb, so can use days and weeks directly.
+        // also assume it maximized months, so we can scale that for months and years.
+
+        let (months, years) = divmod(self.msb.num_months() as UnitSize, 12);
+
+        NuDurationStruct::new(
+            self.base,
+            years,
+            months,
+            self.msb.num_weeks() as UnitSize,
+            self.msb.num_days() as UnitSize,
+            hours,
+            minutes,
+            seconds,
+            fraction,
+        )
+    }
+
+    /// field getters - inverse of [from_b_d_ns].
+    #[inline]
+    pub fn get_base(&self) -> BaseDT {
+        self.base
+    }
+    #[inline]
+    pub fn get_days(&self) -> UnitSize {
+        self.days_from_msb()
+    }
+    #[inline]
+    pub fn get_nanoseconds(&self) -> UnitSize {
+        self.ns_from_lsb()
+    }
+
+    // number of days in msb of duration (from months through days)
+    #[inline]
+    fn days_from_msb(&self) -> UnitSize {
+        (((self.msb.num_months() * 12) + self.msb.num_weeks() * 4) + self.msb.num_days())
+            as UnitSize
+    }
+
+    // number of nanoseconds in lsb of duration (from hours on down)
+    #[inline]
+    fn ns_from_lsb(&self) -> UnitSize {
+        self.lsb
+            .num_nanoseconds()
+            .expect("Overflow getting duration in nanoseconds")
+    }
+}
+
+/*
+impl Display for NuDuration {
+    /// Format inverse of [NuDuration::from_string]
+    todo!();
+}
+*/
+
+/// Serde representation of NuDuration
+#[derive(Debug, Default, PartialEq)]
+pub struct NuDurationStruct {
+    pub base: BaseDT,    // base date/time
+    pub years: UnitSize, // (signed) offsets
+    pub months: UnitSize,
+    pub weeks: UnitSize,
+    pub days: UnitSize,
+    pub hours: UnitSize,
+    pub minutes: UnitSize,
+    pub seconds: UnitSize,
+    pub fraction: UnitSize, // fractional number of nanoseconds (< 1 sec)
```

Comment:
The only thing that has made me question this, so far, is why we don't have milliseconds, microseconds, and nanoseconds but instead of fraction. 🤔

Reply:
This is the utility module supporting the value, the parser will have constants for all the time units.